File: bin/workspace.py
#!/usr/bin/python3
import sys
import os
import json

# 用法：使用执行该脚本并添加left的参数，则i3工作区切换到下一个，否则默认切换到下一个
workspaces = os.popen('i3-msg -t get_workspaces')

try:
    direction = sys.argv[1]
except:
    direction = 'right'

json = json.load(workspaces)
for ws in json:
    if ws['focused'] is True:
        if direction == 'left':
            next_ws = int(ws['name']) - 1
            os.system(f'i3-msg workspace {next_ws}')
            break
        next_ws = int(ws['name']) + 1
        os.system(f'i3-msg workspace {next_ws}')
        break
print("需要left或者right作爲切換參數")

File: bin/workspace.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"strconv"
)

type workspace struct {
	Num     int  `json:"num"`
	Focused bool `json:"focused"`
}

type workspaceList []workspace

var help string = "usage: 使用left或right作为切换参数."

func main() {

	if len(os.Args) != 2 {
		fmt.Fprintf(os.Stdout, "%v\n%v\n", os.Args[0], help)
		return
	}

	arg := os.Args[1]

	data, err := get()
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return
	}

	// 创建workspace的结构体切片数据
	ws := &[]workspace{}
	err = json.Unmarshal(data, ws)

	for _, item := range *ws {
		if item.Focused == true {
			ns := item.Num
			switch arg {
			case "left":
				ns -= 1
			case "right":
				ns += 1
			default:
				fmt.Fprintln(os.Stdout, help)
				return
			}

			r, err := change(ns)
			if err != nil {
				fmt.Fprintln(os.Stdout, err)
				return
			}
			fmt.Fprintln(os.Stdout, string(r))
		}
	}

}

func get() (b []byte, err error) {
	i3cmd := exec.Command("i3-msg", "-t", "get_workspaces")
	// json format but []byte
	cmdOut, err := i3cmd.Output()
	if err != nil {
		return b, err
	}
	return cmdOut, nil
}

func change(n int) (r []byte, err error) {
	ns := strconv.FormatInt(int64(n), 10)
	i3cmd := exec.Command("i3-msg", "workspace", ns)
	r, err = i3cmd.Output()
	if err != nil {
		return r, err
	}
	return r, nil
}
